packages/repository-json-schema/test/integration/build-schema.test.ts

```diff
@@ -0,0 +1,327 @@
+              },
+            },
+          },
+        });
+      });
+
+      it('creates definitions only at the root level of the schema', () => {
+        @model()
+        class CustomTypeFoo {
+          @property() prop: string;
+        }
+
+        @model()
+        class CustomTypeBar {
+          @property.array(CustomTypeFoo) prop: CustomTypeFoo[];
+        }
+
+        @model()
+        class TestModel {
+          @property() cusBar: CustomTypeBar;
+        }
+
+        const jsonSchema = modelToJsonSchema(TestModel);
+        const schemaProps = jsonSchema.properties;
+        const schemaDefs = jsonSchema.definitions;
+        expect(schemaProps).to.deepEqual({
+          cusBar: {
+            $ref: '#definitions/CustomTypeBar',
+          },
+        });
+        expect(schemaDefs).to.deepEqual({
+          CustomTypeFoo: {
+            properties: {
+              prop: {
+                type: 'string',
+              },
+            },
+          },
+          CustomTypeBar: {
+            properties: {
+              prop: {
+                type: 'array',
+                items: {
+                  $ref: '#definitions/CustomTypeFoo',
+                },
+              },
+            },
+          },
+        });
+      });
+    });
+
+    it('properly converts primitive arrays properties', () => {
+      @model()
+      class TestModel {
+        @property.array(Number) numArr: number[];
+      }
+
+      const jsonSchema = modelToJsonSchema(TestModel);
+      expect(jsonSchema.properties).to.deepEqual({
+        numArr: {
+          type: 'array',
+          items: {
+            type: 'number',
+          },
+        },
+      });
+    });
+
+    it('properly converts custom type arrays properties', () => {
+      class CustomType {
+        prop: string;
+      }
+
+      @model()
+      class TestModel {
+        @property.array(CustomType) cusArr: CustomType[];
+      }
+
+      const jsonSchema = modelToJsonSchema(TestModel);
+      expect(jsonSchema.properties).to.deepEqual({
+        cusArr: {
+          type: 'array',
+          items: {
+            $ref: '#definitions/CustomType',
+          },
+        },
+      });
+    });
+
+    it('supports explicit primitive type decoration via strings', () => {
+      @model()
+      class TestModel {
+        @property({type: 'string'})
+        hardStr: Number;
+        @property({type: 'boolean'})
+        hardBool: String;
+        @property({type: 'number'})
+        hardNum: Boolean;
+      }
+
+      const jsonSchema = modelToJsonSchema(TestModel);
+      expect(jsonSchema.properties).to.deepEqual({
+        hardStr: {
+          type: 'string',
+        },
+        hardBool: {
+          type: 'boolean',
+        },
+        hardNum: {
+          type: 'number',
+        },
+      });
+    });
+
+    it('errors out when explicit type decoration is not primitive', () => {
+      @model()
+      class TestModel {
+        @property({type: 'not primitive'})
```

Comment:
Can you remove the space in this type string? Technically, it wouldn't be a valid class/variable reference as a result, so it doesn't quite prove what's being asserted in the `it` statement.
`type: 'NotPrimitive'`